src/time_utils.h

```diff
@@ -10,6 +10,57 @@
 
 #include "export.h"
 
+/* TimescaleDB-specific ranges for valid timestamps and dates: */
+#define TS_EPOCH_DIFF (POSTGRES_EPOCH_JDATE - UNIX_EPOCH_JDATE)
+#define TS_EPOCH_DIFF_MICROSECONDS (TS_EPOCH_DIFF * USECS_PER_DAY)
+
+/* For Timestamps, we need to be able to go from UNIX epoch to POSTGRES epoch
+ * and thus add the difference between the two epochs. This will constrain the
+ * max supported timestamp by the same amount. */
+#define TS_TIMESTAMP_MIN MIN_TIMESTAMP
+#define TS_TIMESTAMP_END (END_TIMESTAMP - TS_EPOCH_DIFF_MICROSECONDS)
+#define TS_TIMESTAMP_INTERNAL_MIN (TS_TIMESTAMP_MIN + TS_EPOCH_DIFF_MICROSECONDS)
+#define TS_TIMESTAMP_INTERNAL_END (TS_TIMESTAMP_END + TS_EPOCH_DIFF_MICROSECONDS)
+
+/* For Dates, we're limited by the timestamp range (since we internally first
+ * convert dates to timestamps). Naturally the TimescaleDB-specific timestamp
+ * limits apply as well. */
+#define TS_DATE_MIN (DATETIME_MIN_JULIAN - POSTGRES_EPOCH_JDATE)
+#define TS_DATE_END (TIMESTAMP_END_JULIAN - POSTGRES_EPOCH_JDATE - TS_EPOCH_DIFF)
+#define TS_DATE_INTERNAL_MIN (TS_TIMESTAMP_MIN + TS_EPOCH_DIFF_MICROSECONDS)
+#define TS_DATE_INTERNAL_END (TS_TIMESTAMP_END + TS_EPOCH_DIFF_MICROSECONDS)
+
+/*
+ * -Infinity and +Infinity in internal (Unix) time.
+ */
+#define TS_TIME_NOBEGIN (PG_INT64_MIN)
+#define TS_TIME_NOEND (PG_INT64_MAX)
+
+#define TS_TIME_IS_INTEGER_TIME(type) (type == INT2OID || type == INT4OID || type == INT8OID)
+#define TS_TIME_DATUM_IS_MIN(timeval, type) (timeval == ts_time_datum_get_min(type))
+#define TS_TIME_DATUM_IS_END(timeval, type) (timeval == ts_time_datum_get_end(type))
```

Comment:
This will return true for both for `TS_TIME_DATUM_IS_MIN(ts_time_datum_get_min(INT4OID), INT4OID)` and TS_TIME_DATUM_IS_MIN(ts_time_datum_get_end(INT4OID), INT4OID)`. I think you need to distinguish the two in some manner or risk hard-to-find bugs.

Reply:
I don't understand. The latter example will return false. Or am I missing something? Can you explain what you mean?

Reply:
Typo, corrected above.

Reply:
I was referring to that there are no checks that the wrong version of the "overloaded" value is passed in.

Reply:
I still don't understand. If they are the same, why wouldn't this return true?

I guess this is related to the other discussed issue about distinct values for these definitions?

Reply:
Yes, it is related to the other discussion. I do not think both should return true.